hail/src/main/scala/is/hail/backend/spark/LowerTableIR.scala

```diff
@@ -1,29 +1,199 @@
 package is.hail.backend.spark
 
-import is.hail.annotations.RegionValue
-import is.hail.backend.Binding
+import java.io.{ByteArrayInputStream, ByteArrayOutputStream}
+
+import is.hail.annotations._
+import is.hail.cxx
 import is.hail.expr.ir._
+import is.hail.expr.types.physical.PTuple
 import is.hail.expr.types.virtual._
+import is.hail.io.CodecSpec
+import is.hail.nativecode._
 import is.hail.rvd.{RVDPartitioner, RVDType}
 import is.hail.utils._
+import org.apache.spark.SparkContext
 import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.Row
 
-case class SparkCollect(
-  child: SparkStage)
+object SparkPipeline {
+  def local(body: IR): SparkPipeline = SparkPipeline(Map.empty[String, SparkStage], body)
+}
+
+case class SparkPipeline(stages: Map[String, SparkStage], body: IR) {
+  def codecSpec: CodecSpec = CodecSpec.defaultUncompressed
+
+  def typ: Type = body.typ
+
+  def execute(sc: SparkContext, region: Region): (PTuple, Long) = {
+    val fields: IndexedSeq[(String, Type)] =
+      stages.map { case (name, stage) => (name, TArray(stage.body.typ)) }.toFastIndexedSeq
+
+    val inputType = TStruct(fields: _*)
+    val rvb = new RegionValueBuilder(region)
+    rvb.start(inputType.physicalType)
+    rvb.startStruct()
+    stages.foreach { case (_, stage) => stage.collect(sc, rvb) }
+    rvb.endStruct()
 
-case class SparkShuffle(
-  child: SparkStage)
+    val ref = Ref(genUID(), inputType)
+    val node = MakeTuple(FastSeq(Subst(body, Env[IR](fields.map { case (name, _) => name -> GetField(ref, name) }.toFastSeq : _*))))
+    val f = cxx.Compile(ref.name, inputType.physicalType, node, optimize = true)
+    (node.pType.asInstanceOf[PTuple], f(region.get(), rvb.end()))
+  }
+}
+
+case class SparkShuffle(child: SparkStage)
+
+case class SparkBinding(name: String, value: SparkPipeline)
 
 case class SparkStage(
-  globals: List[Binding],
+  globals: List[SparkBinding], // globals.head is the globals of the current table; globals.tail can be any other global information that might be needed.
```

Comment:
I really dislike comments when code structure suffices...

Could we do something like:

```scala
case class SparkStage(
    theRealGlobals: SparkBinding,
    otherGlobals: List[SparkBinding],
    ...
) {
    val globals = theRealGlobals +: otherGlobals
```

for both this and the local table ir? This enforces the non-emptiness of `globals` by construction as well as giving us obviously-non-error-causing ways to get the head and tail.